sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetValidation.java

```diff
@@ -0,0 +1,292 @@
+   * 0xDC00-0xDFFF: second block
+   * 0XE000-0xFFFF: single 16 bit block
+   * 
+   * The following UTF code points get mapped into 1 or 2 blocks:
+   * 0x0000 -0xD7FF   (0    -55295)  : 2 bytes, direct mapping
+   * 0xE000 -0xFFFF   (57344-65535)  : 2 bytes, direct mapping
+   * 0x10000-0x10FFFF (65536-1114111): 2 blocks of 2 bytes (not direct..)
+   * 
+   * Note that Unicode 55296-57345 (0xD800 to 0xDFFF) are not used, since it's reserved and used in UTF-16 for the high/low surrogates.
+   * 
+   * We reject 2-byte blocks with 0 (we consider it's binary) even though it's a valid UTF-16 encoding.
+   * 
+   */
+  public boolean isValidUTF16(byte[] buffer) {
+    return isValidUTF16(buffer, false);
+  }
+
+  public boolean isValidUTF16(byte[] buffer, boolean le) {
+    if (buffer.length < 2) {
+      return false;
+    }
+    for (int i = 0; i < buffer.length / 2; i++) {
+      boolean extraByte = false;
+      int c = read16bit(buffer, i, le);
+
+      if (c >= 0xD800 && c < 0xDC00) {
+        // it's a higher surrogate (10 bits)
+        extraByte = true;
+        i++;
+      } else if ((c >= 0xDC00 && c < 0xE000) || c == 0) {
+        return false;
+      }
+      // else it is a simple 2 byte encoding (code points in BMP), and it's valid
+
+      if (extraByte && i < buffer.length / 2) {
+        c = read16bit(buffer, i, le);
+        if (c < 0xDC00 || c >= 0xE000) {
+          // invalid lower surrogate (10 bits)
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Checks if a buffer contains only valid UTF8 encoded bytes. 
+   * It's very effective, giving a clear YES/NO, unless it's ASCII  (unicode < 127), in which case it returns MAYBE.
+   * 
+   * 
+   * First byte:
+   * 0xxxxxxx: only one byte (0-127)
+   * 110xxxxx: 2 bytes       (194-223, as 192/193 are invalid)
+   * 1110xxxx: 3 bytes       (224-239)
+   * 11110xxx: 4 bytes       (240-244)
+   * 
+   * Bytes 2,3 and 4 are always 10xxxxxx (0x80-0xBF or 128-191).
+   * 
+   * So depending on the number of significant bits in the unicode code point, the length will be 1,2,3 or 4 bytes:
+   * 0 -7 bits  (0x0000-007F):  1 byte encoding
+   * 8 -11 bits (0x0080-07FF): 2 bytes encoding
+   * 12-16 bits (0x0800-FFFF): 3 bytes encoding
+   * 17-21 bits (0x10000-10FFFF): 4 bytes encoding
+   */
+  public Result isUTF8(byte[] buffer, boolean rejectNulls) {
+    boolean onlyAscii = true;
+
+    for (int i = 0; i < buffer.length; i++) {
+      byte len;
+      // make it unsigned for the comparisons
+      int c = (0xFF) & buffer[i];
+
+      if (rejectNulls && c == 0) {
+        return Result.newInvalid();
+      }
+      if (c <= 127) {
+        len = 0;
+      } else if (c >= 194 && c <= 223) {
+        len = 1;
+      } else if (c >= 224 && c <= 239) {
+        len = 2;
+      } else if (c >= 240 && c <= 244) {
+        len = 3;
+      } else {
+        return Result.newInvalid();
+      }
+
+      for (i++; len > 0 && i < buffer.length; i++) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)